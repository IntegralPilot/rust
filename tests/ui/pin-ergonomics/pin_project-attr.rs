#![feature(
    pin_ergonomics,
    where_clause_attrs,
    trait_alias,
    extern_types,
    associated_type_defaults,
    fn_delegation,
)]
#![allow(incomplete_features)]
#![pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on crates

// allowed

#[pin_project]
struct Struct {}

#[pin_project]
enum Enum {}

#[pin_project]
union Union {
    field: (),
}

// disallowed

enum Foo<#[pin_project] T, #[pin_project] U = ()> {
    //~^ ERROR `#[pin_project]` attribute cannot be used on function params
    //~| ERROR `#[pin_project]` attribute cannot be used on function params
    #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on enum variants
    UnitVariant,
    TupleVariant(#[pin_project] T), //~ ERROR `#[pin_project]` attribute cannot be used on struct fields
    StructVariant {
        #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on struct fields
        field: U,
    },
}

#[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on traits
trait Trait {
    #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on associated consts
    const ASSOC_CONST: () = ();
    #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on associated types
    type AssocType = ();

    #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on required trait methods
    fn method();
    #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on provided trait methods
    fn method_with_body() {}
}

#[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on trait aliases
trait TraitAlias = Trait;

#[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on inherent impl blocks
impl Struct {
    // FIXME: delegation macros are not tested yet (how to?)
    #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on delegations
    reuse <Struct as std::any::Any>::type_id;

    #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on inherent methods
    fn method() {}
}

#[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on trait impl blocks
impl Trait for Enum {
    #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on trait methods in impl blocks
    fn method() {}
}

#[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on extern crates
extern crate alloc;

#[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on use statements
use std::pin::Pin;

#[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on statics
static STATIC: () = ();

#[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on constants
const CONST: () = ();

#[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on functions
fn f<T, U>(#[pin_project] param: Foo<T, U>)
//~^ ERROR `#[pin_project]` attribute cannot be used on function params
//~| ERROR allow, cfg, cfg_attr, deny, expect, forbid, and warn are the only allowed built-in attributes in function parameters
where
    #[pin_project]
    //~^ ERROR `#[pin_project]` attribute cannot be used on where predicates
    T:,
{
    #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on closures
    || ();
    #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on expressions
    [(), (), ()];
    #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on statements
    let _: Foo<(), ()> = Foo::StructVariant {
        #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on struct fields
        field: (),
    };
    match param {
        #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on match arms
        Foo::UnitVariant => {}
        Foo::TupleVariant(..) => {}
        Foo::StructVariant {
            #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on pattern fields
            field,
        } => {}
    }
}

#[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on modules
mod m {}

#[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on foreign modules
extern "C" {
    #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on foreign types
    type ForeignTy;

    #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on foreign statics
    static EXTERN_STATIC: ();

    #[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on foreign functions
    fn extern_fn();
}

#[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on type alias
type Type = ();

#[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on macro defs
macro_rules! macro_def {
    () => {};
}

#[pin_project] //~ ERROR `#[pin_project]` attribute cannot be used on macro calls
macro_def!();

std::arch::global_asm! {
    "{}",
    #[pin_project] //~ ERROR this attribute is not supported on assembly
    const 0
}

fn main() {}
