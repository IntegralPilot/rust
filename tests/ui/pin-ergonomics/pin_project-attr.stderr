error: `#[pin_project]` attribute cannot be used on macro calls
  --> $DIR/pin_project-attr.rs:135:1
   |
LL | #[pin_project]
   | ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: this attribute is not supported on assembly
  --> $DIR/pin_project-attr.rs:140:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^

error: allow, cfg, cfg_attr, deny, expect, forbid, and warn are the only allowed built-in attributes in function parameters
  --> $DIR/pin_project-attr.rs:84:12
   |
LL | fn f<T, U>(#[pin_project] param: Foo<T, U>)
   |            ^^^^^^^^^^^^^^

error: `#[pin_project]` attribute cannot be used on crates
  --> $DIR/pin_project-attr.rs:10:1
   |
LL | #![pin_project]
   | ^^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on function params
  --> $DIR/pin_project-attr.rs:27:10
   |
LL | enum Foo<#[pin_project] T, #[pin_project] U = ()> {
   |          ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on function params
  --> $DIR/pin_project-attr.rs:27:28
   |
LL | enum Foo<#[pin_project] T, #[pin_project] U = ()> {
   |                            ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on enum variants
  --> $DIR/pin_project-attr.rs:30:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on struct fields
  --> $DIR/pin_project-attr.rs:32:18
   |
LL |     TupleVariant(#[pin_project] T),
   |                  ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on struct fields
  --> $DIR/pin_project-attr.rs:34:9
   |
LL |         #[pin_project]
   |         ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on traits
  --> $DIR/pin_project-attr.rs:39:1
   |
LL | #[pin_project]
   | ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on associated consts
  --> $DIR/pin_project-attr.rs:41:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on associated types
  --> $DIR/pin_project-attr.rs:43:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on required trait methods
  --> $DIR/pin_project-attr.rs:46:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on provided trait methods
  --> $DIR/pin_project-attr.rs:48:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on trait aliases
  --> $DIR/pin_project-attr.rs:52:1
   |
LL | #[pin_project]
   | ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on inherent impl blocks
  --> $DIR/pin_project-attr.rs:55:1
   |
LL | #[pin_project]
   | ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on delegations
  --> $DIR/pin_project-attr.rs:58:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on inherent methods
  --> $DIR/pin_project-attr.rs:61:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on trait impl blocks
  --> $DIR/pin_project-attr.rs:65:1
   |
LL | #[pin_project]
   | ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on trait methods in impl blocks
  --> $DIR/pin_project-attr.rs:67:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on extern crates
  --> $DIR/pin_project-attr.rs:71:1
   |
LL | #[pin_project]
   | ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on use statements
  --> $DIR/pin_project-attr.rs:74:1
   |
LL | #[pin_project]
   | ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on statics
  --> $DIR/pin_project-attr.rs:77:1
   |
LL | #[pin_project]
   | ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on constants
  --> $DIR/pin_project-attr.rs:80:1
   |
LL | #[pin_project]
   | ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on functions
  --> $DIR/pin_project-attr.rs:83:1
   |
LL | #[pin_project]
   | ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on function params
  --> $DIR/pin_project-attr.rs:84:12
   |
LL | fn f<T, U>(#[pin_project] param: Foo<T, U>)
   |            ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on closures
  --> $DIR/pin_project-attr.rs:92:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on expressions
  --> $DIR/pin_project-attr.rs:94:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on struct fields
  --> $DIR/pin_project-attr.rs:98:9
   |
LL |         #[pin_project]
   |         ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on statements
  --> $DIR/pin_project-attr.rs:96:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on match arms
  --> $DIR/pin_project-attr.rs:102:9
   |
LL |         #[pin_project]
   |         ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on pattern fields
  --> $DIR/pin_project-attr.rs:106:13
   |
LL |             #[pin_project]
   |             ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on where predicates
  --> $DIR/pin_project-attr.rs:88:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on modules
  --> $DIR/pin_project-attr.rs:112:1
   |
LL | #[pin_project]
   | ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on foreign modules
  --> $DIR/pin_project-attr.rs:115:1
   |
LL | #[pin_project]
   | ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on foreign types
  --> $DIR/pin_project-attr.rs:117:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on foreign statics
  --> $DIR/pin_project-attr.rs:120:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on foreign functions
  --> $DIR/pin_project-attr.rs:123:5
   |
LL |     #[pin_project]
   |     ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on type aliases
  --> $DIR/pin_project-attr.rs:127:1
   |
LL | #[pin_project]
   | ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: `#[pin_project]` attribute cannot be used on macro defs
  --> $DIR/pin_project-attr.rs:130:1
   |
LL | #[pin_project]
   | ^^^^^^^^^^^^^^
   |
   = help: `#[pin_project]` can be applied to data types and unions

error: aborting due to 40 previous errors

